Code lenses jump to super method/field

kpbochenek opened the pull request.

kpbochenek committed:
Code lenses jump to super method/field

kpbochenek commented on metals/src/main/scala/scala/meta/internal/metals/CodeLensProvider.scala:
```diff
@@ -42,12 +58,11 @@ final class DebugCodeLensProvider(
   ): Seq[l.CodeLens] = {
     semanticdbs.textDocument(path).documentIncludingStale match {
-      case _ if classes.isEmpty =>
+      case None =>
```
need to revert that

kpbochenek replied:
actually this is correct, even if classes.isEmpty we might still need generate lenses for methods

kpbochenek committed:
Extract global table to be created once per code lens request

kpbochenek committed:
Fix searching for SymbolInformation

kpbochenek committed:
Allow jumping to super methods in external dependency code

tgodzik commented on build.sbt:
```diff
@@ -1,3 +1,5 @@
+name := "metals"
```
I don't think we need this change

kpbochenek replied:
when a project is opened in intellij now it gets gibberish name, with this change it will be displayed in a list of projects as 'metals' and it doesn't change anything in regard to functionality

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/ImplementationProvider.scala:
```diff
@@ -2,8 +2,13 @@ package scala.meta.internal.implementation
 import org.eclipse.lsp4j.TextDocumentPositionParams
-import scala.meta.internal.mtags.Semanticdbs
-import scala.meta.internal.mtags.{Symbol => MSymbol}
+import scala.meta.internal.mtags.{
```
Let's keep single line imports

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/ImplementationProvider.scala:
```diff
@@ -345,19 +356,32 @@ final class ImplementationProvider(
     findSemanticDbForSymbol(symbol).flatMap(findSymbol(_, symbol))
   }
 
+  def symbolDef(symbol: String): Option[SymbolDefinition] = {
```
symbolDef -> symbolDefinition

Let's use full names.

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/MethodImplementation.scala:
```diff
@@ -197,6 +197,15 @@ object MethodImplementation {
     }
   }
 
+  def checkSignaturesEqual(
```
Because of the generics comparing signatures is not symetric, we need parent -> child

kpbochenek replied:
This was already implemented, I didn't change anything here except exposing this method without a need to provide context. It calls `signaturesEqual`

kpbochenek replied:
ah ok parameter names, you can mark multiple lines doing comment, it helps to understand what you mean better.

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/ImplementationProvider.scala:
```diff
@@ -345,19 +356,32 @@ final class ImplementationProvider(
+    for {
+      symbolDefinition <- symbolDef(symbol)
+      document <- findSemanticdb(symbolDefinition.path)
+    } yield TextDocumentWithPath(document, symbolDefinition.path)
```
path is in the document already as URI. Granted it's only relative, but if needed it should be easy to get it.

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/SuperMethodProvider.scala:
```diff
@@ -0,0 +1,211 @@
+) {
+
+  def findSuperForMethodOrField(
+      msi: SymbolInformation,
```
`msi` -> `methodSymbolInformation`?

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/SuperMethodProvider.scala:
```diff
@@ -0,0 +1,211 @@
+    } else {
+      skip += si
+      si.signature match {
+        case clz: ClassSignature =>
```
`clz` - lets use better names

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/SuperMethodProvider.scala:
```diff
@@ -0,0 +1,211 @@
+                "scala/AnyVal#"
+              ).contains(si.symbol)
+            )
+            .filterNot(p => skip.exists(_.symbol == p.symbol))
```
We can join the two filters

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/SuperMethodProvider.scala:
```diff
@@ -0,0 +1,211 @@
+          val semDB = findSemanticDbWithPathForSymbol(si.symbol)
+          (si, semDB) +: expParents
+        case sig: TypeSignature =>
+          findSymbol(sig.lowerBound.asInstanceOf[TypeRef].symbol)
```
We can reuse the dealias function from the `ImplementationProvider`

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/SuperMethodProvider.scala:
```diff
@@ -0,0 +1,211 @@
+              ).contains(si.symbol)
+            )
+            .filterNot(p => skip.exists(_.symbol == p.symbol))
+          val mainParents = parents.headOption
```
Why a separate case for the first parent? To linearize we should first go up the first parent and then others.

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/SuperMethodProvider.scala:
```diff
@@ -0,0 +1,211 @@
+            .toList
+            .flatten
+          val withParents = if (parents.nonEmpty) {
+            parents.tail.reverse.flatMap { linParent =>
```
I don't think this should be a reverse

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/CodeLensProvider.scala:
```diff
@@ -70,22 +103,60 @@ final class DebugCodeLensProvider(
     }
   }
+
+  private def createSuperMethodCommand(
```
Let's have a separate code lens providers and just join their results in CodeLensProvider object?

We can have them both take SemanticDB as input. In a similar way as CodeActions

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/codeactions/GoToSuperMethod.scala:
```diff
@@ -0,0 +1,115 @@
+package scala.meta.internal.metals.codeactions
```
This is not a code action, it should be probably in the implementation package.

tgodzik commented on tests/unit/src/test/scala/tests/SuperMethodLspSuite.scala:
```diff
@@ -0,0 +1,411 @@
+  test("simple") {
+    val code =
+      """
+        |package a
```
The test could look like that:
```
        |trait A { def <<1>>xxx: String = "A" }
        |trait B extends A { override def <<2->1>>xxx: String = "B -> " + super.xxx }
        |trait C extends A { override def <<3->2>>xxx: String = "C -> " + super.xxx }
        |trait D extends C { override def <<4->3>>xxx: String = "D -> " + super.xxx }
```
you wouldn't need the separate map.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1430,6 +1464,9 @@ class MetalsLanguageServer(
 
   private def quickConnectToBuildServer(): Future[BuildChange] = {
+
+    import scala.meta.internal.metals.codeactions.GoToSuperMethod
```
I don't think this is needed

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -1151,11 +1174,14 @@ class MetalsLanguageServer(
+      codeLensSync(params).asJava
     }
 
+  def codeLensSync(params: CodeLensParams): List[CodeLens] = {
```
Why does it need to be a separate function?

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:
```diff
@@ -383,6 +384,28 @@ class MetalsLanguageServer(
+      implementationProvider.findSemanticDbWithPathForSymbol
+    )
+
+    goToSuperMethod = new GoToSuperMethod(
```
Can we keep `SuperMethodProvider` and `GoToSuperMethod` together? It's not really a separate functionality.

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/CodeLensProvider.scala:
```diff
@@ -70,22 +103,60 @@ final class DebugCodeLensProvider(
 object CodeLensProvider {
   import scala.meta.internal.metals.JsonParser._
 
+  type LensGoSuperCache =
```
This is not a good place to keep the cache. The cache needs to be dropped on each semanticDB update, so that everything stays up to date and let's not cache entire TextDocuments :fearful: 

What we would really need is just an information if a symbol has a parent, GOTO Parent will calculate the rest.

kpbochenek replied:
this is per request cache ...
If you have a class X with 20 methods you don't want for each method to calculate hierarchy of X.

tgodzik commented on metals/src/main/scala/scala/meta/internal/implementation/SuperMethodProvider.scala:
```diff
@@ -0,0 +1,211 @@
+    result
+  }
+
+  def findSuperForMethodOrFieldChecked(
```
You should use `MethodImplementation.findParentSymbol`, no need to reimplement it all.

kpbochenek replied:
I don't think I could use that as `MethodImplementation.findParentSymbol` doesn't care about linearization and does not necessarily finds next parent in linearized hierarchy. 
We can discuss that if you are convinced it is usable in this context.

lukaszwawrzyk committed:
Add support for package prefix

kpbochenek committed:
Changes after review

kpbochenek committed:
Super method hierarchy support

kpbochenek committed:
Cosmetic changes

kpbochenek committed:
Implement missing tests

tgodzik committed:
Update scalameta to 4.3.4 and fix existing tests

kpbochenek committed:
Fixes after review

kpbochenek committed:
Scalafix comply and fix imports

kpbochenek committed:
Add requested testcase

Commit:
Add helper to pre-download Metals dependencies.

Previously, there was no convenient way to build a Docker image with all
of Metals dependencies pre-cached. This commit adds a new
`DownloadDependencies` main entrypoint that downloads artifacts that
Metals is likely to depend on.

We can't guarantee Metals will always work offline after running this
method. For example, the following cases are not handled:

* user opens Metals in a Gradle/Maven/Mill/sbt build.
* user has configured an older Scalafmt version.

A separate tool is needed to populate those cases (at least for now)

Context: https://github.com/eclipse/che-plugin-registry/pull/356

Commit:
Use consistent download method for mdoc, mtags and SemanticDB

ckipp01 committed:
only check relevant text for default indent

kpbochenek committed:
SuperMethodProvider as private field

olafurpg committed:
Merge pull request #1501 from olafurpg/download-deps

Add helper to pre-download Metals dependencies.

kpbochenek committed:
Comply with new scalafix

tgodzik committed:
Merge pull request #1499 from tgodzik/update-scalameta

Update scalameta to 4.3.4 and fix existing tests

lukaszwawrzyk committed:
Generate bloop.settings.json with settings refresh

olafurpg committed:
Merge pull request #1470 from lukaszwawrzyk/package-prefix

Export packagePrefix to bloop

tgodzik committed:
Do not include scala boot library, since it may lead to wrong library being indexed

tgodzik committed:
Add a separate rambo symbol search for the rambo presentation compiler

ckipp01 committed:
Merge pull request #1505 from ckipp01/multiline

Only check relevant text for default indent

tgodzik committed:
Add support for Scala 2.12.11

kpbochenek committed:
Fixes after review

kpbochenek committed:
Fix empty line imports

kpbochenek committed:
Remove another asInstanceOf

tgodzik committed:
Merge pull request #1510 from tgodzik/update-scalameta-2.12.11

Add support for Scala 2.12.11

Commit:
Use a custom IDEA launcher under a new flag.

Previously, it was not possible to export a Pants build and launch
IntelliJ with a custom IntelliJ version (for example, 2020.1 EAP).
Now, the `--intellij-launcher APP` flag allows users to pass in a custom
launcher.

olafurpg committed:
Merge pull request #1513 from olafurpg/idea-app

Use a custom IDEA launcher under a new flag.

tgodzik committed:
Merge pull request #1507 from tgodzik/fix-boot-classpath

Do not include scala boot library

tgodzik committed:
Add release notes for v0.8.2 and bump versions

kpbochenek committed:
Get rid of asInstanceOf[MethodSignature]

tgodzik committed:
Merge pull request #1515 from tgodzik/0.8.2-release

Add release notes for v0.8.2 and bump versions

lukaszwawrzyk committed:
Reworks

kpbochenek committed:
Refactor symbol search after review

lukaszwawrzyk committed:
Ensure .bloop dir exists

Commit:
Always launch IntelliJ in the project's root directory

Commit da56189c introduced a hack, that was applicable for IntelliJ
2019.3. It required to open the root's parent directory instead of the
root directory, except the first time, when the project was imported.

In IntelliJ it's safe to always open the root directory. What is more,
opening the parent directory couses that, the project is not
identified as BSP-based.

olafurpg committed:
Merge pull request #1506 from lukaszwawrzyk/refresh-project

Generate bloop.settings.json with project refresh command

olafurpg committed:
Merge pull request #1516 from tpasternak/always-open-intellij-in-root-dir

Always launch IntelliJ in the project's root directory

Commit:
Undeprecate 2.12.10.

Currently, users get the following warning

> You are using a legacy Scala version 2.12.10, which might not be
> supported in future versions of Metals. Please upgrade to Scala 2.12.11.

Scala 2.12.11 just came out, we should wait at least a few months before
calling it "legacy".

Commit:
Prepare for a new release

tgodzik committed:
Merge pull request #1517 from olafurpg/non-deprecate

Undeprecate 2.12.10.

tgodzik committed:
Add more information to release notes and bump other versions

tgodzik committed:
Merge pull request #1518 from olafurpg/release

Prepare for a new release

lukaszwawrzyk committed:
Don't run bloop install when opening intellij

olafurpg committed:
Merge pull request #1519 from lukaszwawrzyk/double-refresh-fix

Don't run bloop install when opening intellij

Commit:
Minor polish for Pants integration

* Refine message if the Bloop project is not yet exported.
* Generate symbolic links for directories to improve the "project files" view

olafurpg committed:
Merge pull request #1521 from olafurpg/no-targets

Minor polish for Pants integration

ckipp01 committed:
update lsp4j to enable more 3.15 support

ckipp01 committed:
add in another apply for HoverMarkup and update worksheet hover tests

ckipp01 committed:
update documentation for Decoration Protocol

This is to go along with the update of lsp4j in order to account for LSP 3.15

gabro committed:
Add "case class" option in NewFilesProvider

ckipp01 committed:
Merge pull request #1522 from ckipp01/lsp4j

Update lsp4j to enable more 3.15 support

gabro committed:
Add CreateNewFile code action

gabro committed:
Merge pull request #1525 from gabro/case-class-pick

Add "case class" option in NewFilesProvider

gabro committed:
Fix indents

gabro committed:
Add one more test

gabro committed:
Handle blank file name

gabro committed:
s/Create new.../Create new symbol '$name'...

gabro committed:
s/CreateNewFile/CreateNewSymbol/g

gabro committed:
Fix race condition

gabro committed:
Merge pull request #1528 from gabro/new-file-action

Add "Create new symbol '$name'..." code action

Duhemm committed:
Export `tags` in `BloopPants`

`tags` is a new field of the Bloop configuration, which is used to
propagate more information about the target described by a given
configuration file.

At the moment, this field is used to indicate that a given target is a
test target. All other targets are tagged with `library`.

olafurpg committed:
Merge pull request #1532 from Duhemm/fastpass-tags

Export `tags` in `BloopPants`

kpbochenek committed:
super method code lenses disabled by default

Commit:
Add one more version to list of known old Bloop versions

olafurpg committed:
Merge pull request #1533 from olafurpg/old-bloop

Add one more version to list of known old Bloop versions
